preserve aten::to device in export training

avikchaudhuri opened the pull request:
Summary:
With training IR, we cannot rely on trapping `to()` in `FunctionalTensor` because the regular decomposition kicks it first, and that can cause it to be optimized away.

So instead we preserve it until we functionalize, and then replace it explicitly with `_to_copy()`.

Test Plan: expected test failures go away

Differential Revision: D61883878

tugsbayasgalan commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -161,7 +162,7 @@ def inner(*flat_args):
         # It doesn't matter if we run this under predispatch or not because it is
         # only for figuring out metadata
-        mode = FunctionalTensorMode(_allow_token_discovery=True)
+        mode = FunctionalTensorMode(_allow_token_discovery=True, export=is_export)
```
Not sure why we need this extra flag? The previous code path and the new code path should be same, so i feel we don't need this extra flag?

avikchaudhuri replied:
`FunctionalTensorMode` doesn't get `export` set otherwise, and I don't want to rewrite the op in environments other than `export`.

tugsbayasgalan replied:
So there are actually two FunctionalTensorMode inside aot_autograd, one for collecting metadata and one for actually tracing. For the one that is for metadata collection, we actually don't need it i think.

avikchaudhuri replied:
If I instead try `_detect_infra_mode(torch._C._TorchDispatchModeKey.FUNCTIONAL)` from inside `FunctionTensorMode.__torch_dispatch__` I get `None`. The previous code path looked for `.export` in the result of this call.

avikchaudhuri replied:
> So there are actually two FunctionalTensorMode inside aot_autograd, one for collecting metadata and one for actually tracing. For the one that is for metadata collection, we actually don't need it i think.

So the problem is, once we've changed the registration we have to ensure that every time we hit this op we replace it.

This is the error I was getting otherwise (even in the metadata collection pass, before actual tracing):
```
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/scripts/avik/pt2/example.py", line 4524, in test_device_to_static
    ep_ = ep_.run_decompositions(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/export/exported_program.py", line 97, in wrapper
    return fn(*args, **kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/export/exported_program.py", line 1062, in run_decompositions
    return _decompose_exported_program(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/export/exported_program.py", line 678, in _decompose_exported_program
    gm, new_graph_signature = _decompose_and_get_gm_with_new_signature_constants(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/export/exported_program.py", line 433, in _decompose_and_get_gm_with_new_signature_constants
    aten_export_artifact = _export_to_aten_ir(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/export/_trace.py", line 632, in _export_to_aten_ir
    gm, graph_signature = transform(aot_export_module)(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_functorch/aot_autograd.py", line 1248, in aot_export_module
    fx_g, metadata, in_spec, out_spec = _aot_export_function(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_functorch/aot_autograd.py", line 1482, in _aot_export_function
    fx_g, meta = create_aot_dispatcher_function(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_functorch/aot_autograd.py", line 522, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_functorch/aot_autograd.py", line 623, in _create_aot_dispatcher_function
    fw_metadata = run_functionalized_fw_and_collect_metadata(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_functorch/_aot_autograd/collect_metadata_analysis.py", line 169, in inner
    flat_f_outs = f(*flat_f_args)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_functorch/_aot_autograd/utils.py", line 182, in flat_fn
    tree_out = fn(*args, **kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_functorch/_aot_autograd/traced_function_transforms.py", line 859, in functional_call
    out = PropagateUnbackedSymInts(mod).run(
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/fx/interpreter.py", line 146, in run
    self.env[node] = self.run_node(node)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/fx/experimental/symbolic_shapes.py", line 5480, in run_node
    result = super().run_node(n)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/fx/interpreter.py", line 203, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/fx/interpreter.py", line 275, in call_function
    return target(*args, **kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_ops.py", line 717, in __call__
    return self._op(*args, **kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_higher_order_ops/utils.py", line 64, in inner
    return autograd_not_implemented_inner(op, deferred_error, *args, **kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_higher_order_ops/utils.py", line 37, in autograd_not_implemented_inner
    result = operator(*args, **kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_ops.py", line 717, in __call__
    return self._op(*args, **kwargs)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/torch/_subclasses/functional_tensor.py", line 505, in __torch_dispatch__
    outs_unwrapped = func._op_dk(
RuntimeError: Found a custom (non-ATen) operator whose output has alias annotations: aten::to.dtype_layout(Tensor(a) self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a). We only support functionalizing operators whose outputs do not have alias annotations (e.g. 'Tensor(a)' is a Tensor with an alias annotation whereas 'Tensor' is a Tensor without. The '(a)' is the alias annotation). The alias annotation specifies that the output Tensor shares storage with an input that has the same annotation. Please check if (1) the output needs to be an output (if not, don't return it), (2) if the output doesn't share storage with any inputs, then delete the alias annotation. (3) if the output indeed shares storage with an input, then add a .clone() before returning it to prevent storage sharing and then delete the alias annotation. Otherwise, please file an issue on GitHub.

While executing %to : [num_users=1] = call_function[target=torch.ops.aten.to.dtype_layout](args = (%x,), kwargs = {dtype: torch.int64, layout: torch.strided, device: cpu})
Original traceback:
File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/344b081a93d75e32/scripts/avik/pt2/__example__/example#link-tree/scripts/avik/pt2/example.py", line 4494, in forward
    return x.to("cpu")
```

Good news is that there's another way to fix this without adding a flag to the metadata collection pass, which is to *not* change the registration to `NotImplemented` (but do everything else in `_override_composite_implicit_decomp`). Turns out I have to do this anyway when investigating the other test cases around `float()`. So I'll change this in an updated PR.

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -3463,8 +3459,6 @@ def forward(self, x):
-    @testing.expectedFailureTrainingIRToRunDecompNonStrict
     def test_device_to_mutation(self):
         class Module(torch.nn.Module):
             def forward(self, x):
```
There should be 2, 3 more instance of this task in this test.

avikchaudhuri replied:
Oh, you mean `test_float_conversion` and `test_device_to_mutation_float`? Yeah I missed those.

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -413,7 +420,11 @@ def _decompose_and_get_gm_with_new_signature_constants(
-        with fake_mode:
+        with fake_mode, _override_composite_implicit_decomp(
+            # preserve understanding it is not functional, replace with aten._to_copy later
+            (torch.ops.aten.to.dtype_layout,),
```
This is more of a question to @bdhirsh and @zou3519: We are basically disabling CIA decomp for aten.to here. By the time, we hit the post-dispatch FunctionalTensor mode in python, do we guarantee that it has never been went through C++ functionalization decomp? Do we always make sure we hit the PythonKey before hitting Functionalize key in post-dispatch world? I think previously it didn't matter because we only preserved functiuonal CIA ops anyways.

tugsbayasgalan commented on torch/export/exported_program.py:
```diff
@@ -413,7 +423,14 @@ def _decompose_and_get_gm_with_new_signature_constants(
+            # Preserve variants of aten::to understanding that they are mutating/aliasing
+            # and their CompositeImplicitAutograd kernels will not become NotImplemented.
+            # We will later replace them with aten._to_copy when functionalizing.
+            (torch.ops.aten.to.dtype_layout, torch.ops.aten.to.dtype),
```
Could you actually make it a seperate context manager? In my PR, I am merging this path so that it always calls export_to_aten_ir and it already takes many flags as function call.

avikchaudhuri replied:
I'll create an alias.

tugsbayasgalan commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -342,6 +342,19 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
             kwargs = {}
 
+        mode = _detect_infra_mode(torch._C._TorchDispatchModeKey.FUNCTIONAL)
+        if self.export or (mode is not None and mode.export):
```
I think it is safe to assert that mode here is a self actually.

avikchaudhuri committed:
preserve aten::to device in export training (#134622)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134622

With training IR, we cannot rely on trapping `to()` in `FunctionalTensor` because the regular decomposition kicks it first, and that can cause it to be optimized away.

So instead we preserve it until we functionalize, and then replace it explicitly with `_to_copy()`.

Test Plan: expected test failures go away

Reviewed By: tugsbayasgalan

Differential Revision: D61883878

bdhirsh commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -342,6 +342,23 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         if kwargs is None:
             kwargs = {}
 
+        if self.export:
```
mentioned to @tugsbayasgalan offline, but I would much rather have this logic be populated into export's decomposition table than have it get hardcoded directly into `FunctionalTensorMode`.
